crates/apub/src/api/test.rs

```diff
@@ -0,0 +1,42 @@
+use activitypub_federation::config::Data;
+use lemmy_api_common::context::LemmyContext;
+use lemmy_db_schema::{
+  source::{
+    instance::Instance,
+    local_user::{LocalUser, LocalUserInsertForm},
+    person::{Person, PersonInsertForm},
+  },
+  traits::Crud,
+};
+use lemmy_db_views::structs::LocalUserView;
+use lemmy_utils::error::LemmyResult;
+
+#[derive(Default)]
+pub struct TestUser {
+  pub name: String,
+  pub bio: Option<String>,
+  pub admin: bool,
+}
+
+impl TestUser {
```

Comment:
probably makes sense to have it over there anyway so it can be used by other tests as well

Reply:
wouldn't changing this to an `impl` for `LocalUserView` prevent using the `TestUser` struct with defaults for the creation? how would this have to be written?

Reply:
You can use whatever params you like to create the LocalUserView. There's no need to have a special struct for a test user, especially if all you're checking is 1 or 2 fields.

```rust
impl LocalUserView {
  pub async fn create_test_user(self, pool, ..whatever_other_params, ) -> LemmyResult<Self> {
    let instance_id = Instance::
    ...
  }
}
```

Reply:
which in turn means there is no `default()`, and all params must always be provided, making this less usable again.
the function currently creates a `Person` and then a `LocalUser`. there are many places in the code where tests are setting more than just the name, bio and admin status. for this to work properly for all those configurations it doesn't make sense to build a list of 20+ arguments for `LocaluserView::default().create()`.
if it's not that generic there is no point in moving it over to `LocalUserView` in the first place, as most tests will have different requirements in what needs to be set on the user.

even in just the two files touched in this PR, this already sets the `admin` flag in one test and the `bio` in the other test. looking at `crates/db_schema/src/impls/local_user.rs`, this sets the email of the `LocalUser`. `crates/db_views/src/post_view.rs` sets the `bot_account` flag of the `Person`.

if this is moved to a more central location with the intention of calling it from other tests it'll have to be able to take all the parameters used by the different tests. i don't intend to add support for more parameters than these two in this PR to keep the PR more readable, but i could create another PR afterwards that will change other tests to also use the same common test user creation function.

Reply:
do you have a recommendation for how to structure this around `LocalUserView` or in another more central place?

Reply:
@dessalines @Nutomic ping?

Reply:
Dessalines is out this week, he should be back on Monday. Lets wait for him to have a look again.

Reply:
You don't need to provide all the params, I'm not sure what's confusing.

```rust
impl LocalUserView { 
  pub async fn create_test_user(self, pool, INSERT_ALL_PARAMS_YOU_NEED_HERE ) -> LemmyResult<Self> { 
    // Move things from line 21 into here
  } 
}
```

Reply:
Yes you do, see my previous comment.

If the function is supposed to be usable in all the different tests across different crates and modules it'll have to accept all the different inputs they require, otherwise it's not really improving the situation if you still have to update the person/local_user after the creation to get it in the state needed in those tests.

The overall idea is to reduce boilerplate and duplicate code, which should also help avoid mistakes when creating objects during testing and improve readability by reducing the lines of code required in tests.

The intention of the change in this PR was to use a common method to both create a `Person` and also create the associated `LocalUser`. When you use `SomeInsertForm`, you have the option of writing it like this:

```rs
let local_user_form = LocalUserInsertForm {
  admin: Some(true),
  ..LocalUserInsertForm::test_form(inserted_person.id)
};
```

This however does not allow you to simplify the creating of a `Person` _and_ a `LocalUser` at once for tests, as you will always have to explicitly create both forms.
I don't think it makes sense to put a bunch of unrelated refactoring into this PR, so I'm also fine with just reverting this to having just a local creation function in the test module of each of the affected files in here.

For the actual refactoring, which would be in a separate PR, what do you think about adding a `submit(pool)` method to forms as first step in that direction?

```rs
impl PersonInsertForm {
  pub async fn submit(&self, pool: &mut DbPool<'_>) -> Result<Person, Error> {
    Person::create(pool, self).await
  }
}
```

See https://github.com/Nothing4You/lemmy/commit/3ed0879ec2381b324d8304a730244351c1ba597d for what this could look like in practice.
Do you think it would make sense to refactor this for all the forms and also use it outside of the tests?

Reply:
`.submit(pool)` does look much cleaner than the current code.

Reply:
submit might be a pointless wrapper, and maybe a little confusing. For example, for the `PostSaved` form, we have a `Saveable` trait, so you can do save and unsave (one which inserts, one which deletes the row) Adding `submit` wouldn't be clear which one it needs to do.

@Nothing4You here's what i mean, you can eliminate the pointless testuser and all those new imports by just moving the test user creation to the object impl itself: https://github.com/Nothing4You/lemmy/pull/5

Reply:
Of course there should only be one way to run a specific query, not multiple.

Regarding your `create_test_user()`, the problem is that every caller needs to specify bio and admin, even if they are completely irrelevant. It still works in this example because there are only two extra params, but if its used across the whole codebase there would be many more params which are only needed a single time. Then its much easier to omit unused params using struct init with `..Default::default()`.

Though it would be better to make a separate PR for refactoring only.